harness/determined/pytorch/_pytorch_context.py

```diff
@@ -57,4 +82,224 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+        if self.hvd_config.use:
+
+            def filter_named_parameters(
+                model: nn.Module, optimizer: torch.optim.Optimizer,  # type: ignore
+            ) -> List:
+                """
+                filter_named_parameters filters the named parameters of a specified optimizer out
+                of all the named parameters from a specified model. We need this function because
+                a torch.optim.Optimizer doesn't store parameter names and we need the names of
+                the parameters when mapping parameters to each horovod.DistributedOptimizer.
+                """
+                opt_params = {
+                    p for group in optimizer.param_groups for p in group.get("params", [])
+                }
+                return [(name, p) for name, p in model.named_parameters() if p in opt_params]
+
+            use_compression = self.hvd_config.fp16_compression
+            optimizer = hvd.DistributedOptimizer(
+                optimizer,
+                named_parameters=filter_named_parameters(cast(nn.Module, self.model), optimizer),
+                backward_passes_per_step=self.hvd_config.aggregation_frequency,
+                compression=hvd.Compression.fp16 if use_compression else hvd.Compression.none,
+            )
+            logging.debug("Initialized optimizer for distributed and optimized parallel training.")
+
+        self.optimizers.append(optimizer)
+        return optimizer
+
+    def _LRScheduler(
+        self,
+        lr_scheduler: torch.optim.lr_scheduler._LRScheduler,
+        step_mode: pytorch.LRScheduler.StepMode,
+    ) -> pytorch.LRScheduler:
+        # Initialize LR schedulers. This block must happen after wrapping the optimizers using
+        # horovod.DistributedOptimizer because torch.optim._LRScheduler directly access the
+        # optimizer parameter groups, which are also Horovod.DistributedOptimizer relies on.
+        wrapped = pytorch.LRScheduler(lr_scheduler, step_mode)
+        self.lr_schedulers.append(wrapped)
+        return wrapped
+
+    def _init_device(self) -> None:
+        self.n_gpus = len(self.env.container_gpus)
+        if self.hvd_config.use:
+            check.gt(self.n_gpus, 0)
+            # We launch a horovod process per GPU. Each process
+            # needs to bind to a unique GPU.
+            self.device = torch.device(hvd.local_rank())
+            torch.cuda.set_device(self.device)
+        elif self.n_gpus > 0:
+            self.device = torch.device("cuda", 0)
+        else:
+            self.device = torch.device("cpu")
+        check.is_not_none(self.device)
+
+    def _to_device(self, data: pytorch._Data) -> pytorch.TorchData:
+        return pytorch.to_device(
+            data, self.device, self.warning_logged[_WarningLogs.FAILED_MOVING_TO_DEVICE]
+        )
+
+    def _use_amp(self) -> bool:
+        return self.mixed_precision != "O0"
+
+    def _configure_amp(
+        self,
+        model: torch.nn.Module,
```

Comment:
non-blocking: We should consider supporting the same options as amp.initialize here.  @aaron276h already mentioned exposing a bunch of options, but there two other important differences:
  - apex.amp.initialize() allows you to pass in a single model or a list of models
  - apex.amp.initialize() allows you to pass in None for the optimizer

https://github.com/NVIDIA/apex/blob/master/apex/amp/frontend.py#L195

I marked this as non-blocking because I don't think it needs to block *this* PR, but I would block on this before exposing this to the API public.

Reply:
Since we are supporting multiple models via sub-models of the main model, I think we should only let users pass in only one model. But we should let users pass in at least one optimizer since `apex` supports None only for inference.

Reply:
I have added the arguments of `amp.initialize` to `_configure_amp`. So will resolve this conv.

Reply:
(Unresolving this, please do not resolve a thread before I have a chance to see your response)

> Since we are supporting multiple models via sub-models of the main model

Is this true just for right now, or is it true in the state of the final API as well?  I thought our planned API worked just fine with multiple separate models.

Reply:
It's true for the final API as well. Implementing multiple sub-models would be very simple for users and implementing multiple models would be hacky for us. Pytorch `nn.Module` sees a member field `self.sub_model=model_a` as a first-class citizen compared to `self.sub_model[0]=model_a`. The latter can't be recognized automatically by the `nn.module` implementation for serialization, deserialization, parameters, buffers, and etc.

Reply:
We have had a discussion for multiple sub-models API and it seems fine for internal users. https://determined-ai.slack.com/archives/C0147BBR61E/p1591306980029400

Reply:
Discussed offline. Will go with multiple models.